lib/bap/bap_self.ml

```diff
@@ -60,4 +61,94 @@ module Create() = struct
   let warning f = message Warning ~section:name f
   let error f = message Error ~section:name f
 
+  module Param = struct
+    let plugin_name = name
+
+    type 'a t = 'a ref
+    type 'a parser = string -> [ `Ok of 'a | `Error of string ]
+    type 'a printer = Format.formatter -> 'a -> unit
+    type 'a converter = 'a parser * 'a printer
+
+    let main = ref Term.(const ())
+
+    let conf_file_options : (string, string) List.Assoc.t =
+      let conf_filename =
+        let (/) = Filename.concat in
+        Bap_config.confdir / plugin_name / "config" in
+      let string_splitter str =
+        let str = String.strip str in
+        match String.split str ~on:'=' with
+        | k :: _ when String.prefix k 1 = "#" -> None
+        | [""] | [] -> None
+        | [k] -> invalid_argf
+                   "Maybe comment out \"%s\" using # in config file?" k ()
+        | k :: vs -> Some (k, String.concat ~sep:"=" vs) in
+      let split_filter = List.filter_map ~f:string_splitter in
+      try
+        In_channel.with_file
+          conf_filename ~f:(fun ch -> In_channel.input_lines ch
+                                      |> split_filter)
+      with Sys_error _ -> []
+
+    let get_from_conf_file name =
+      List.Assoc.find conf_file_options ~equal:String.Caseless.equal name
+
+    let get_from_env name =
+      let name = "BAP_" ^ String.uppercase (plugin_name ^ "_" ^ name) in
+      try
+        Some (Sys.getenv name)
+      with Not_found -> None
+
+    let get_param ~(converter:'a converter) ~default ~name =
+      let value = default in
+      let str = get_from_conf_file name in
+      let str = match get_from_env name with
+        | Some _ as v -> v
+        | None -> str in
+      let parse str =
+        let parse, _ = converter in
+        match parse str with
+        | `Error err
+          -> invalid_argf
+               "Could not parse \"%s\" for name \"%s\"" str name ()
+        | `Ok v -> v in
+      let value = match str with
+        | Some v -> parse v
+        | None -> value in
+      ref value
+
+    let create converter ~default ?(docv="VAL") ~doc ~name : 'a t =
```

Comment:
I am going to rename all the functions to their proper names once we decide on them. All the current names can be considered placeholders, just to not get confused (as offline, we were switching between too many names, so I thought it'd be better to just keep one standard and then once we are done deciding, we can change in one go) :)